py3status/modules/google_calendar.py

```diff
@@ -0,0 +1,489 @@
+                curr_event_format,
+                {'summary': event_str['summary'],
+                 'location': event_str['location'],
+                 'description': event_str['description'],
+                 'start_time': event_str['start_time'],
+                 'end_time': event_str['end_time'],
+                 'start_date': event_str['start_date'],
+                 'end_date': event_str['end_date'],
+                 'time_to': time_to_composite})
+
+            self.py3.composite_update(event_formatted, {'index': index})
+            responses.append(event_formatted)
+
+            index += 1
+
+        format_separator = self.py3.safe_format(self.format_separator)
+        self.py3.composite_update(format_separator, {'index': 'sep'})
+        responses = self.py3.composite_join(format_separator, responses)
+
+        return {'events': responses}
+
+    def google_calendar(self):
+        """
+        The method that outputs the response.
+
+        First, we check credential authorization. If no authorization, we display an error message,
+        and try authorizing again in 5 seconds.
+
+        Otherwise, we fetch the events, build the response, and output the resulting composite.
+        """
+        if not self.is_authorized:
+            self.is_authorized = self._authorize_credentials()
+
+            return {'cached_until': self.py3.time_in(self.cache_timeout),
+                    'full_text': self.py3.safe_format(self.format)}
+        else:
+            if not self.no_update:
+                self.events = self._get_events()
+
+            return {
+                'cached_until': self.py3.time_in(self.cache_timeout),
+                'composite': self.py3.safe_format(self.format, self._build_response())
+            }
+
+    def on_click(self, event):
+        if self.is_authorized and self.events is not None:
+            """
+            If button_refresh is clicked, we allow the events to be updated if the
+            last event update occured at least 1 second ago. This prevents a bug that
+            can crash py3status since refreshing the module too fast results in
+            incomplete event information being fetched as _get_events() is called
+            repeatedly.
+
+            Otherwise, we disable event updates.
+            """
+            try:
+                self.no_update = True
+
+                self.button = event['button']
+                self.button_index = event['index']
+
+                if self.button == self.button_toggle:
+                    if self.button_index != 'sep':
+                        self.button_states[self.button_index] = \
+                            not self.button_states[self.button_index]
+                elif self.button == self.button_refresh:
+                    now = datetime.datetime.now()
+                    diff = (now - self.last_update).seconds
+                    if diff > 1:
+                        self.no_update = False
```

Comment:
Explain what's happening here?

Reply:
So this is the fix for that py3status crash bug I was talking about. To see what I'm talking about we have to distinguish between two separate things:

(1) Refreshing the module
(2) Fetching events 

Doing (1) as often as we want is no problem. However, if (2) is done successively very quickly, the module will crash. So I have to decouple (1) from (2), meaning if  we press `button_open` or `button_toggle`, we want to refresh the module but NOT fetch events. Why? Because if you hit say `button_toggle` multiple times in a second without decoupling the two, this will try to fetch events over and over and eventually py3status will crash. Nothing gets logged when this happens, but I was able to determine that it has to do with trying to fetch events again before the last fetch was incomplete...so refreshing events too fast is bad. 

So the solution I did has two elements:

One, we don't update events for `button_toggle` or `button_open`. This is what `self.no_update` is for. We set it to True at the start of `on_click()` so that events can't be fetched again until the button event is handled. This prevents fetching the events too often by repeated presses of a button. Perhaps `self.no_update` isn't a good name since it's not very accurate. Perhaps `self.no_event_update` or `self.no_event_fetch` would be clearer.

Two, if `button_refresh` is pushed we want to do (1) and (2), since we want to fetch events and then refresh the module to reflect any changes. However, this brings the crash problem back. If we fetch events too often in a short period of time (by pressing `button_refresh` many times quickly), the module will crash. So what I do to fix this case is, I use `self.last_update` in `_get_events()` to save the last time the module was updated. When `button_refresh` is pressed, we check how long it's been since the last event fetch. If it's been more than 1 second since the last time we fetched events, we allow them to be fetched again. This keeps the events from being fetched too often, which causes the crash.

I think your suggestion for handling url opening is right. It would make more sense to handle it in `on_click()`, but it won't remove the need for `self.no_update` because we still need to decouple refreshing the module from fetching events.

Reply:
>your suggestion for handling url opening

Also, I forget... See `gcal` as it's doing the `self.url` too... and see `mpris` and `timer` as both modules are working with `index` too... and could be useful to you.

>Perhaps self.no_update isn't a good name since it's not very accurate.

Will `self.clicking` makes more sense in this context? If not, `self.no_update` is fine.

>it won't remove the need for self.no_update

I use same thing in `gcal` with `self.scrolling`.

Reply:
Ok, I'll have a look at those modules, thanks. We have no need for `self.no_update` when using `button_refresh` since we want to update events, so `self.clicking` probably doesn't make sense since we only care about it for 2 out of 3 buttons. 

I just had a look at gcal. It looks like `_organize()` is where we create individual events from `gcalcli` , output and you don't do this if we're moving the focus marker (i.e. scrolling)? If so, it's the same idea, yeah.